Chia-Hao Lee, 2021.03.07

## General question:
### When we calculate standard deviation, how many significant figures we should keep?
Example: s = std(a) = 0.123456789..., 

s = 0.1  ?

s = 0.12 ?

s = 0.123?

In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [14]:
loc = 330.5
scale = 8.123456789
num = 34000
iterations = 10000
size = (num, iterations)
a = np.random.normal(loc, scale, size)

print('Gaussian mean at', loc)
print('Gaussian sigma = ', scale)
print('Gaussian size = ', size)

Gaussian mean at 330.5
Gaussian sigma =  8.123456789
Gaussian size =  (34000, 10000)


In [15]:
std_list = []
mean_list = []
k = iterations/10
for i in range(iterations):
    if i%int(k)==0:
        print('%s th std = '%(i), np.std(a[:,i]))
    std_list.append(np.std(a[:,i]))
    mean_list.append(np.mean(a[:,i]))
print('STD of the  %s stds = %s, each std was calculated from %s numbers' %(iterations, np.std(std_list), num))
print('The STD can be approximated by std/sqrt(2*num) = %s /sqrt(2*%s) = %s' %(np.mean(std_list), num, np.mean(std_list)/(2*num)**0.5))

0 th std =  8.09452668010568
1000 th std =  8.132975818819968
2000 th std =  8.118279886063785
3000 th std =  8.068658803776664
4000 th std =  8.104785725702753
5000 th std =  8.08392118721399
6000 th std =  8.095580249985641
7000 th std =  8.08976878781442
8000 th std =  8.121497192396104
9000 th std =  8.059549428237489
STD of the  10000 stds = 0.030977465797876694, each std was calculated from 34000 numbers
The STD can be approximated by std/sqrt(2*num) = 8.1235310357041 /sqrt(2*34000) = 0.031152319451000318


## Note
We generated a lot of random numbers from a Gaussian distribution, we can separate them into different groups. Each group has (#num) values, and we have (#iterations) groups.

Everytime we calculate std from a different group, the std will be slightly different because we only have limited samples(#num). We can calculate these std for multiple (#iterations), and we'll see the STD for these stds also changes slowly as we increase (#iterations). 

The reason that std fluctuates is the same as why STD fluctuate, they're both originates from statistical error of "sampling".

### As you may notice, this implies that all the estimate from a distribution has its own uncertainty, and you need infinite samples to determine all of them.

In [16]:
print('In this example, ')
print('The mean of first %s samples = %s' %(num , mean_list[0]))
print('The std from these %s samples = %s' %(num, std_list[0]))
print('However, we know that the fluctuations between %s stds = %s' %(iterations, np.std(std_list)))

print('The std can be ranging from %s to %s' %(np.mean(std_list)-np.std(std_list),np.mean(std_list)+np.std(std_list)))
print("\nThere's NO point to claim the std with more figures than its own fluctuation.")
print("Same idea, there's NO point to claim the mean with more figures than its own fluctuation.")
print("The fluctuation of the mean is the 'standard error of the mean', expected to be ", std_list[0]/np.sqrt(num) ,
      "\nwhich is std/sqrt(N), not std(standard deviation for all the samples) ")
print("The SEM from these simulation is ", np.std(mean_list))

print('\nFor std = 8.xxxx, we do not need to report the extra digits because it does not help us to get a better mean.')
print('In order to claim the std below decimal point, we need 0.5*(8/0.01)^2 = 320000 samples......')

In this example, 
The mean of first 34000 samples = 330.5187560263953
The std from these 34000 samples = 8.09452668010568
However, we know that the fluctuations between 10000 stds = 0.030977465797876694
The std can be ranging from 8.092553569906224 to 8.154508501501978

There's NO point to claim the std with more figures than its own fluctuation.
Same idea, there's NO point to claim the mean with more figures than its own fluctuation.
The fluctuation of the mean is the 'standard error of the mean', expected to be  0.043898734463516303 
which is std/sqrt(N), not std(standard deviation for all the samples) 
The SEM from these simulation is  0.043743475396619436

For std = 8.xxxx, we do not need to report the extra digits because it does not help us to get a better mean.
In order to claim the std below decimal point, we need 0.5*(8/0.01)^2 = 320000 samples......


### In short, 
1. You did 34000 measurements and want to report the distribution of raw measurement: 

d = 331 $\pm$ 8

2. You want to report the possible spread of the mean you got:

d_mean = 330.5 $\pm$ 0.04
